File: #2_Add_Two_Numbers/Solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} l1
    # @param {ListNode} l2
    # @return {ListNode}
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode(-1)
        cur, iter1, iter2 = dummy, l1, l2
        sum = 0
        while iter1 or iter2 or sum == 1:
            if iter1:
            	sum += iter1.val
            	iter1 = iter1.next
            if iter2:
            	sum += iter2.val
            	iter2 = iter2.next
            cur.next = ListNode(sum % 10)
            cur = cur.next
            sum /= 10
        return dummy.next


n1 = ListNode(2)
n2 = ListNode(4)
n3 = ListNode(3)
m1 = ListNode(5)
m2 = ListNode(6)
m3 = ListNode(4)
n1.next = n2
n2.next = n3
m1.next = m2
m2.next = m3
sol = Solution()
re = sol.addTwoNumbers(n1, m1)
while re != None:
	print re.val
	re = re.next

File: #2_Add_Two_Numbers/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;
        ListNode cur1 = l1; ListNode cur2 = l2;
        int rem = 0;
        while (cur1 != null && cur2 != null) {
        	int val = cur1.val + cur2.val + rem;
        	ListNode nd = new ListNode(val % 10);
        	rem = val / 10;
        	prev.next = nd;
        	prev = nd;
        	cur1 = cur1.next;
        	cur2 = cur2.next;
        }
        while (cur1 != null) {
        	int val = cur1.val + rem;
        	ListNode nd = new ListNode(val % 10);
        	rem = val / 10;
        	prev.next = nd;
        	prev = nd;
        	cur1 = cur1.next;
        }
        while (cur2 != null) {
        	int val = cur2.val + rem;
        	ListNode nd = new ListNode(val % 10);
        	rem = val / 10;
        	prev.next = nd;
        	prev = nd;
        	cur2 = cur2.next;
        }
        if (rem > 0) {
        	ListNode nd = new ListNode(rem);
        	prev.next = nd;
        }
        return dummy.next;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	ListNode h1 = new ListNode(2);
    	ListNode n2 = new ListNode(4);
    	ListNode n3 = new ListNode(3);
    	h1.next = n2;
    	n2.next = n3;
    	ListNode h2 = new ListNode(5);
    	ListNode n4 = new ListNode(6);
    	ListNode n5 = new ListNode(4);
    	h2.next = n4;
    	n4.next = n5;
    	ListNode r = sol.addTwoNumbers(h1, h2);
    	while (r != null) {
    		System.out.print(r.val + " ");
    		r = r.next;
    	}
    	System.out.println();
    }
}